### DISCUSSION TITLE
Software debounce missing input events

### DISCUSSION THREAD

**Discussion by thagrol** [gpiozero/gpiozero#937] [Q&A]
Having created a gpiozero.Button object and set call backs for for when_held, when_pressed, and when_released if bounce_time is anything other than the default events are being randomly missed.

This occurs even when bounce time is set as low as 0.001

Code sees presses but not releases, releases without presses, holds without presses, holds without releases etc.

Hardware was a single tactile switch connected between GPIO 19 and ground. Pi Zero and a Pi2 showed the same behaviour.

**Comment by lurch** [comment]
I don't have time to look into this right now, but I wonder if it might be another side-effect of the issues discussed in #385 ?

See also #378

**Comment by DaddyTheRunner** [comment]
I was experimenting with the `bounce_time` parameter and a simple tactile switch and got some odd behavior.  Here's my code:

```python
from gpiozero import RGBLED, Button
from time import sleep

rgb = RGBLED(25, 12, 16)
red_button = Button(6, bounce_time=1.0)

rgb.off()

red = (1,0,0)
green = (0,1,0)
delay = 0.5

## Signal ready to run
rgb.color = red
sleep(1)
rgb.color = green

## Testing button bounce
try:
    count = 0
    
    while True:
        red_button.wait_for_press()
        count += 1
        print(count)

except KeyboardInterrupt:
    # Clean up
    rgb.off()

```

When I ran the code with `bounce_time = None`, I got behavior I expected.  Typically a single momentary press would generate a stream of numbers between 1200 - 2500 and then stop when I released the button.

However, when I set `bounce_time` to any floating point value, the stream of numbers never stopped if I held the button down for less than the `bounce_time` until I interrupted the program with `ctrl-C`.  If I held the button down for a period of time greater than `bounce_time`, the stream of numbers stopped when I released the button.

Unless I don't understand how the `bounce_time` works (which is quite possible), it would appear that the `button.wait_for_press` function doesn't get "reset" until after the button has been held longer than `bounce_time` and then released.  That is not the behavior I would expect.

**Comment by 1deadpixl** [comment]
Hi, all.

I'm wondering if there is an update on this issue since it's been a while. I'm having the same problem using v1.4.1 installed with pip on a RPi 3B. It looks like the timing problem noted in issue #385 was fixed, but this debounce problem is still around. Any idea what the root cause is and if/when a fix will be released? It seems to be a rather popular issue given that this page is the second result when searching 'gpiozero debounce'.

I just started my first project using gpiozero (have used RPi.GPIO in the past) and things are so much easier. I'm especially in love with the sources/values mechanism. It would be great to see this debounce thing get worked out though since it's a critical component to building a nice physical interface.

Thanks for all the hard work!

**Comment by rdehuyss** [comment]
Unusable because of this...

**Comment by bennuttall** [comment]
Can anyone compare gpiozero code with RPi.GPIO and show that it has a difference debounce?

gpiozero is using RPi.GPIO underneath (by default). If you change the pin factory you can test it using other libraries, like pigpio. https://gpiozero.readthedocs.io/en/stable/api_pins.html#changing-the-pin-factory

**Reply by waveform80** [reply]
> Can anyone compare gpiozero code with RPi.GPIO and show that it has a difference debounce?

There won't be any difference there because we just used whatever debounce / glitch / bounce filter the underlying library provides. Here's the chunk where we activate event detection in RPi.GPIO just passing along the bounce value we're given (scaled to ms): https://github.com/gpiozero/gpiozero/blob/master/gpiozero/pins/rpigpio.py#L231

Now, there's definitely some mis-understandings of various things in this thread, so here's a quick debunk section:

* The various `wait_for_X` methods are *level* triggered, not *edge* triggered. For example, if a Button is not held down and you `wait_for_press`, then it'll only return when someone presses the button. However, if a Button is *already* held down and you `wait_for_press` it'll immediately return *even though the button hasn't changed state* because it's just waiting for the press *level*, not a press *edge*. This is in contrast to the various `when_X` events which are indeed *edge* triggered (both exist because both are useful in different scenarios).
* Software debounce solutions (unlike hardware solutions which are generally rather more reliable) usually just involve ignoring edges for some period of time after an initial edge. I can't recall if that's how RPi.GPIO's operates, but it certainly sounds like it from various descriptions here.
* I do know that pigpio's [glitch filter](http://abyz.me.uk/rpi/pigpio/cif.html#gpioGlitchFilter) is rather more sophisticated than this. It samples a given line periodically and only returns level changes that persist longer than the provided period. This is probably what people *think* most software debounce is like, but it's extremely rare because it's expensive computationally (compared to just ignoring stuff for a given period), but pigpio has some clever tricks for reducing that cost.
* As in other cases, we just pass values down to the selected driver scaled appropriately. In other words, if you're finding RPi.GPIO's debounce iffy, you may have better results simply [switching gpiozero's pin factory](https://gpiozero.readthedocs.io/en/stable/api_pins.html#changing-the-pin-factory) over to pigpio which'll then use its rather more fancy glitch filter instead.

**Comment by lurch** [comment]
@bennuttall I suspect this might be another side-effect of https://github.com/RPi-Distro/python-gpiozero/issues/697#issuecomment-463413405 i.e. RPi.GPIO missing rising/falling edge events?

**Comment by thagrol** [comment]
Looking into this now but I'm hitting what appears to be a major issue with the NativeFactory on a Pi2: the entire Pi hangs when creating a Button object with no error message.

Will post a seperate issue with more details shortly.

**Comment by thagrol** [comment]
OK, after some testing with alternate pin factories here are my findings:

RPIO: couldn't get the module to import
PiGPIO: appears to work correctly. A brief test did not exhibit this issue
Native: hung the entire Pi
Mock: not tested

I was using python 2 on a Pi 2 running raspbian stretch.
Stratch my results for the native pin factory. I think something unrelated is causing the hang.

**Comment by sailingbo** [comment]
Hi-  Just wanted to see if there have been any updates here.  I'm using a simple footpedal as button(17) and when bounce_time=0, it functions almost as expected.  All presses/releases/helds are accounted for.  However, occasionally, the footpedal will bounce and register 2 presses/releases when I only performed (or meant to perform) 1.  

As I understand it, preventing this is the whole point of the debounce function.  If I set button(17, bounce_time=0.1) or anything at all other than 0, the button becomes essentially non-functioning.   Only 1 out of every 10-15 presses/holds are registered.  I haven't found a correlation to the length of time I wait between presses/holds.  It just becomes unusable.

Any ideas?  Thanks!

**Comment by bennuttall** [comment]
@sailingbo see my last comment https://github.com/gpiozero/gpiozero/issues/550#issuecomment-500743781 - come input on this would be helpful

**Reply by blackestberry** [reply]
Hi, 
as a non-expert, I hope I'm at the right place. I still have this problem with my Raspberry Pi 2 Model B. 
Is there a solution yet? 

As I expanded my system functionality, I realized that `when_released` often misses the real-world-event. This makes it quite "inconvenient" to set e.g. `hold_repeat` to True (-> non-stopping event loop, even if the button has already been released).

I tested different external pull-up resistor values and different buttons.
I've updated to V1.6.2 (By uninstalling and reinstalling, as the standard update command says, V1.5.0 is the latest version [edit: Stretch]). Unfortunately no improvement.

**Best workaround** is still to set `bounce_time` to its default value and risk multiple events at once sometimes. 

Thank you for your great lib!!
